Implement hsplit, vsplit, dsplit

muthuArivoli opened the pull request:
Related to #38349

muthuArivoli committed:
add basic implementation

muthuArivoli commented:
@mruberry Here is what I have so far. The problem I'm running into is that torch.split actually uses the word `tensor` as its argument, while hsplit is using `input` as its argument. Here is what I mean:
```
>>> torch.split(tensor=torch.tensor([1,2,3]), split_size_or_sections=2)
(tensor([1, 2]), tensor([3]))
>>> torch.hsplit(tensor=torch.tensor([1,2,3]), split_size_or_sections=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: hsplit() received an invalid combination of arguments - got (split_size_or_sections=int, tensor=Tensor, ), but expected one of:
 * (Tensor input, tuple of ints split_size_or_sections)
      didn't match because some of the keywords were incorrect: tensor
 * (Tensor input, int split_size_or_sections)
      didn't match because some of the keywords were incorrect: tensor

>>> torch.split(input=torch.tensor([1,2,3]), split_size_or_sections=2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: split() got an unexpected keyword argument 'input'
>>> torch.hsplit(input=torch.tensor([1,2,3]), split_size_or_sections=2)
(tensor([1, 2]), tensor([3]))
```
I can't be sure, but it could be something to do with https://github.com/muthuArivoli/pytorch/blob/6fc2dfade150fb1b2e0b58900b77d8ef1914e070/torch/tensor.py#L447

mruberry commented:
Is the issue just that there's a positional argument name inconsistency? I would just implement dsplit, hsplit, and vsplit as you are with the modern names and file an issue after the PR is in to track the inconsistency. I don't think it's a blocking issue.

muthuArivoli committed:
add test and docs

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-splits

muthuArivoli committed:
fixes

muthuArivoli committed:
fix xla

muthuArivoli commented:
@mruberry PTAL. Sorry for the delay, I had a lot of midterms and projects over the past two weeks, so I wasn't able to work on this.

Alright, so other than the argument naming being different, these functions also differ from split in another way. For the second argument, `split_size_or_sections`, torch.split can accept a tensor, while these functions cannot (they will throw an error). This doesn't seem to be documented well for torch.split, and I'm also not sure if the way that torch.split handles tensors is always the most intuitive. For instance, `torch.split(x, [2])` is actually not equal to `torch.split(x,torch.tensor([2]))`. Let me know if I need to support tensors for the second argument.

Also, side note, thanks for the suggestions on the linalg issue, I'll look at them and pick one up. I do have an open PR that ports lstsq to Aten, but it's still waiting for a review. Should I link you in the PR for a review?

mruberry commented:
> @mruberry PTAL. Sorry for the delay, I had a lot of midterms and projects over the past two weeks, so I wasn't able to work on this.

Of course. Midterms and projects sound like important priorities. We appreciate your work and whatever time you have.

> Alright, so other than the argument naming being different, these functions also differ from split in another way. For the second argument, `split_size_or_sections`, torch.split can accept a tensor, while these functions cannot (they will throw an error). This doesn't seem to be documented well for torch.split, and I'm also not sure if the way that torch.split handles tensors is always the most intuitive. For instance, `torch.split(x, [2])` is actually not equal to `torch.split(x,torch.tensor([2]))`. Let me know if I need to support tensors for the second argument.

Since this has been started, @kurtamohler started implementing a NumPy-compatible tensor_split (see https://github.com/pytorch/pytorch/pull/45168). hsplit in NumPy is actually implemented using split, not array_split, however:

https://github.com/numpy/numpy/blob/291ac7f21e59440d249d0afaed3213a79cf70535/numpy/lib/shape_base.py#L882

This is consistent with its documentation. The difference between array_split and split in NumPy is very small, however: it's just an additional error check on the divisibility of the relevant dimension.

What would you think about waiting (not long) for #45168 to go in and then using that to implement these functions? This is a different approach than I originally recommended, and eventually we'll want to update these functions to remove the few lines of extra check code and call a NumPy-compatible split directly, but I think this will simplify the implementation and prevent us from having to update the actual behavior of these functions later, too. 

@kurtamohler, what do you think of that approach? 

> Also, side note, thanks for the suggestions on the linalg issue, I'll look at them and pick one up. I do have an open PR that ports lstsq to ATen, but it's still waiting for a review. Should I link you in the PR for a review?

Yes, please! Sorry it's been languishing. Sometimes we get a notification and then 100 more pile up over it and it's lost.

kurtamohler commented:
Just to make sure I understand the argument correctly: Since `h/v/dsplit` are using `split` and `split_with_sizes`, their behavior will be different than numpy `h/v/dsplit` (or will take some extra work to make their output match numpy). If we use `tensor_split` underneath instead, behavior will match numpy behavior (except for the extra divisibility error check that Mike mentioned). Furthermore, it will be simple to update the implementation of `h/v/dsplit` once we have a numpy-like `split` function.

If that's all correct, then I agree with that plan.

In addition, it would be good to have unit tests that compare results from `torch.h/v/dsplit` and `numpy.h/v/dsplit`.

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-splits

muthuArivoli committed:
use tensor split

muthuArivoli committed:
Merge remote-tracking branch 'upstream/master' into implement-splits

muthuArivoli committed:
fix tests

muthuArivoli commented:
@mruberry PTAL. I've updated this to use tensor_split.

kurtamohler commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1125,6 +1125,28 @@ std::vector<Tensor> unsafe_split(const Tensor& self, int64_t split_size, int64_t
 
+std::vector<Tensor> hsplit(const Tensor& self, int64_t split_size) {
+  TORCH_CHECK(self.dim() >= 1, "hsplit requires a tensor with at least 1 dimension")
+  if (self.dim() > 1) {
+    TORCH_CHECK(self.sizes()[1] % split_size == 0, "tensor split does not result in an equal division")
+    return at::tensor_split(self, split_size, 1);
+  }
+  TORCH_CHECK(self.sizes()[0] % split_size == 0, "tensor split does not result in an equal division")
+  return at::tensor_split(self, split_size, 0);
```
```suggestion
  int64_t dim = (self.dim() == 1) ? 0 : 1;
  TORCH_CHECK(self.sizes()[dim] % split_size == 0, "tensor split does not result in an equal division")
  return at::tensor_split(self, split_size, dim);
```

kurtamohler replied:
Also, there might be a better way to tell the user what they can do to avoid the error. Maybe the error message could be something like `"self.size(", dim, ") is not divisible by split_size"`.

mruberry replied:
I like @kurtamohler's suggestion. Might as well populate the values, too with something like "self.size(dim)=<value> is not divisible by split_size=<split_size".

kurtamohler commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3230,6 +3230,31 @@
   variants: function, method
   device_guard: False
 
+- func: hsplit.int(Tensor(a) self, int indices_or_sections) -> Tensor(a)[]
```
Could we rename all of these `indices_or_sections` variables to either `indices` (when it's `int[]`) or `sections` (when it's int)? Also, the functions in `aten/src/ATen/native/TensorShape.cpp` should match these argument names.

kurtamohler replied:
Note, I think `indices_or_sections` is still a good name for the Python interface.

kurtamohler commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1155,6 +1177,24 @@ std::vector<Tensor> unsafe_split_with_sizes(const Tensor& self, IntArrayRef spli
 
+std::vector<Tensor> hsplit(const Tensor& self, IntArrayRef split_sizes) {
+  TORCH_CHECK(self.dim() >= 1, "hsplit requires a tensor with at least 1 dimension")
+  if (self.dim() > 1) {
+    return at::tensor_split(self, split_sizes, 1);
+  }
+  return at::tensor_split(self, split_sizes, 0);
```
```suggestion
  return at::tensor_split(self, split_sizes, (self.dim() == 1) ? 0 : 1);
```

kurtamohler commented on torch/_torch_docs.py:
```diff
@@ -1324,6 +1324,136 @@ def merge_dicts(*dicts):
+
+The input tensor is required to have at least one dimension. 
+
+This is equivalent to calling :func:`torch.tensor_split` with :attr:`dim=1` for 
```
Shouldn't this be dim=0 for 1D and dim=1 for greater than 1D?

mruberry replied:
"This is equivalent to..." -> "If :attr:\`input\` is one dimensional this is equivalent to calling torch.tensor_split(input, indices_or_sections, dim=0) (the split dimension is zero), and if :attr:\`input\` has two or more dimensions it's equivalent to calling torch.tensor_split(input, indices_or_sections, dim=1) (the split dimension is 1), except that if :attr:\`indices_or_sections\` is an integer it must evenly divide the split dimension or a runtime error will be thrown."

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1324,6 +1324,136 @@ def merge_dicts(*dicts):
+           r"""
+hsplit(input, indices_or_sections) -> List of Tensors
+
+Split a tensor into multiple tensors horizontally (column wise).
```
"Split" -> "Splits"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1324,6 +1324,136 @@ def merge_dicts(*dicts):
+
+Split a tensor into multiple tensors horizontally (column wise).
+
+The input tensor is required to have at least one dimension. 
```
"input" -> ":attr:\`input\`"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1324,6 +1324,136 @@ def merge_dicts(*dicts):
+           r"""
+vsplit(input, indices_or_sections) -> List of Tensors
+
+Split a tensor into multiple tensors vertically (row wise).
```
Suggestions above can be applied here and to dsplit, too.

mruberry commented on test/test_torch.py:
```diff
@@ -19093,6 +19093,75 @@ def verify_against_numpy(t):
                           [0, 0]], device=device, dtype=dtype)
         verify_against_numpy(t)
 
+    def _test_special_splits(self, torch_fn, np_fn, dim_range, dim, device, dtype):
```
Can we also add some method_test entries so the autograd and jit functionality is tested, too?

mruberry replied:
Would you also add a test (or three) to TestViewOps? It's a test suite in test_torch.py that verifies functions like this are properly returning views. @kurtamohler I actually neglected to mention this for tensor_split and we should add a test for it there, too. These tests are typically very simple. For example:

```
def test_reshape_view(self, device):
        t = torch.ones(5, 5, device=device)
        v = torch.reshape(t, (25,))
        self.assertTrue(self.is_view_of(t, v))

        v[6] = 0
        self.assertEqual(t[1, 1], v[6])
```

mruberry reviewed (commented):
This PR still looks great, @muthuArivoli, and I'm excited to make these three functions available to PyTorch users!

I made a couple suggestions in the tests and docs. Thank you, @kurtamohler, for also reviewing. There's also some bookkeeping to do (see https://github.com/pytorch/pytorch/pull/45168/files for references):

- test/test_autograd.py can be updated to put these ops in the complex allow list, like tensor_split
- tools/autograd/gen_autograd.py should add these functions to its `RETURNS_VIEWS_OF_INPUT` list
- torch/testing/_internal/common_methods_invocations.py needs entries for these functions added (see the tensor_split entries for an example)
- I think test/test_jit.py's `EXCLUDE_SCRIPT_AD_CHECK` will also need an update, analogous to the updates for tensor_split

PyTorch is a complicated (and multifaceted) system, and I really appreciate your attention to detail on this PR, @muthuArivoli.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3230,6 +3230,31 @@
+  use_c10_dispatcher: full
+  variants: function, method
+
+- func: vsplit.int(Tensor(a) self, int indices_or_sections) -> Tensor(a)[]
```
Add a third signature corresponding to tensor_split where indices_or_sections can be a tensor
